File: src/logger.ts
import { appendFileSync, existsSync, mkdirSync } from "node:fs";
import { join } from "node:path";
import type { WorldState, TurnResult, WorldEvent } from "./types.js";

let logsDir = "logs";

export function initLogger(dir: string): void {
  logsDir = dir;
  if (!existsSync(logsDir)) {
    mkdirSync(logsDir, { recursive: true });
  }
}

export function logTurnResult(result: TurnResult): void {
  const line = JSON.stringify(result) + "\n";
  appendFileSync(join(logsDir, "turns.jsonl"), line);
}

export function logEvent(event: WorldEvent): void {
  const line = JSON.stringify(event) + "\n";
  appendFileSync(join(logsDir, "events.jsonl"), line);
}

export function printTurnSummary(
  world: WorldState,
  results: TurnResult[],
): void {
  const alive = world.agents.filter((a) => a.alive);
  const totalEnergy = alive.reduce((sum, a) => sum + a.energy, 0);

  console.log(
    `\n[T${String(world.turn).padStart(3, "0")}] ` +
      `pop=${alive.length} totalEnergy=${totalEnergy}`,
  );

  for (const r of results) {
    const parts: string[] = [];
    if (r.transfer) {
      parts.push(`TRANSFER ${r.transfer.amount} → ${r.transfer.to}`);
    }
    const outputPreview = r.rawOutput
      .replace(/\n/g, " ")
      .slice(0, 60);
    parts.push(`"${outputPreview}..."`);

    const status = r.energyAfter <= 0 ? " DEAD" : "";
    const invokerTag = world.agents.find((a) => a.id === r.agentId)?.invoker ?? "?";
    console.log(
      `  ${r.agentName}[${invokerTag}]: E=${r.energyBefore}→${r.energyAfter}${status} ${parts.join(", ")}`,
    );
  }
}

File: src/logger.py
import json
import os
from dataclasses import asdict

from .types import RoundResult, WorldEvent, WorldState


_logs_dir: str = "logs"


def init_logger(logs_dir: str) -> None:
    global _logs_dir
    _logs_dir = logs_dir
    os.makedirs(logs_dir, exist_ok=True)


def log_round_result(result: RoundResult) -> None:
    path = os.path.join(_logs_dir, "rounds.jsonl")
    with open(path, "a") as f:
        f.write(json.dumps(asdict(result)) + "\n")


def log_event(event: WorldEvent) -> None:
    path = os.path.join(_logs_dir, "events.jsonl")
    with open(path, "a") as f:
        f.write(json.dumps(asdict(event)) + "\n")


def print_round_summary(world: WorldState, results: list[RoundResult]) -> None:
    alive = [a for a in world.agents if a.alive]
    transfers = [r for r in results if r.commands.transfer]
    sends = sum(len(r.commands.sends) for r in results)

    print(f"\n--- Round {world.round} ---")
    print(f"  Population: {len(alive)}/{len(world.agents)}")
    if transfers:
        print(f"  Transfers: {len(transfers)}")
    if sends:
        print(f"  Messages: {sends}")
    for r in results:
        agent = next((a for a in world.agents if a.id == r.agent_id), None)
        if agent is None:
            continue
        status = "ALIVE" if agent.alive else "DEAD"
        line = f"  {r.agent_name}: E={r.energy_before:.2f}->{r.energy_after:.2f} [{status}]"
        if r.commands.transfer:
            line += f" (gave {r.commands.transfer.amount} to {r.commands.transfer.to})"
        if r.commands.sends:
            line += f" ({len(r.commands.sends)} msg)"
        print(line)
